### PR TITLE
Telegram webhook

### PR DESCRIPTION
**Pull request by techknowlogick**
As title

### CODE REVIEW

**Review by sapk — COMMENTED** [review]
Could you add some tests ?

**Comment by techknowlogick** [comment]
@sapk updated tests

**Comment by techknowlogick** [comment]
@sapk / @axifive do you have time for a review of this PR? If not that's ok, no rush.

**Comment by lunny** [comment]
When created an issue, the webhook failed:
```json
{"ok":false,"error_code":400,"description":"Bad Request: can't parse entities: Unsupported start tag \"!--\" at byte offset 154"}
```

**Comment by techknowlogick** [comment]
thanks for testing @lunny. Can you post the the request that was made to telegram, I wonder if I need to escape anything.

**Comment by lunny** [comment]
headers
```
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Content-Length,Content-Type,Date,Server,Connection
Connection: keep-alive
Content-Length: 129
Content-Type: application/json
Date: Wed, 15 Aug 2018 04:09:49 GMT
Server: nginx/1.12.2
```
body
```json
{
  "text": "[\u003ca href=\"http://localhost:3000/lunny/git\"\u003elunny/git\u003c/a\u003e] Issue opened: \u003ca href=\"http://localhost:3000/api/v1/repos/lunny/git/issues/4\"\u003e#4 new issue\u003c/a\u003e\n\n\u003c!--\r\n    1. Please speak English, this is the language all of us can speak and write.\r\n    2. Please ask questions or configuration/deploy problems on our Discord \r\n       server (https://discord.gg/NsatcWJ) or forum (https://discourse.gitea.io).\r\n    3. Please take a moment to check that your issue doesn't already exist.\r\n    4. Please give all relevant information below for bug reports, because \r\n       incomplete details will be handled as an invalid report.\r\n--\u003e\r\n",
  "parse_mode": "HTML"
}
```

**Comment by techknowlogick** [comment]
@lunny from my testing it looks like Telegram doesn't like HTML comments `<!-- -->`. So we have two options, webhooks fail with error message that you get above, or I strip out HTML comments from all content before it gets sent to telegram. Do you have a preference?

**Comment by techknowlogick** [comment]
@lunny I've just adding in escaping for HTML comments. Please try again.

**Comment by lafriks** [comment]
@lunny can you check if this is ok now?

**Review by lafriks — APPROVED** [review]
Did not test but from code part looks ok to me :)

File: models/webhook_telegram.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+	return &TelegramPayload{
+		Message: title,
+	}, nil
+}
+
+func getTelegramDeletePayload(p *api.DeletePayload) (*TelegramPayload, error) {
+	// created tag/branch
+	refName := git.RefEndName(p.Ref)
+	title := fmt.Sprintf(`[<a href="%s">%s</a>] %s <a href="%s">%s</a> created`, p.Repo.HTMLURL, p.Repo.FullName, p.RefType,
```
Comment (jonasfranz) [inline models/webhook_telegram.go]: Replace `created` with `deleted` in code and comment.

File: templates/repo/settings/webhook/telegram.tmpl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{{if eq .HookType "telegram"}}
+	<p>{{.i18n.Tr "repo.settings.add_telegram_hook_desc" "https://core.telegram.org/bots" | Str2html}}</p>
+	<form class="ui form" action="{{.BaseLink}}/settings/hooks/telegram/{{if .PageIsSettingsHooksNew}}new{{else}}{{.Webhook.ID}}{{end}}" method="post">
+		{{.CsrfTokenHtml}}
+		<div class="required field {{if .Err_BotToken}}error{{end}}">
+			<label for="bot_token">{{.i18n.Tr "repo.settings.bot_token"}}</label>
+			<input id="bot_token" name="bot_token" type="text" value="{{.TelegramHook.BotToken}}" autofocus required>
+		</div>
+    <div class="required field {{if .Err_ChatID}}error{{end}}">
+      <label for="chat_id">{{.i18n.Tr "repo.settings.chat_id"}}</label>
```
Comment (jonasfranz) [inline templates/repo/settings/webhook/telegram.tmpl:10]: Translation missing

File: templates/repo/settings/webhook/telegram.tmpl [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{{if eq .HookType "telegram"}}
+	<p>{{.i18n.Tr "repo.settings.add_telegram_hook_desc" "https://core.telegram.org/bots" | Str2html}}</p>
+	<form class="ui form" action="{{.BaseLink}}/settings/hooks/telegram/{{if .PageIsSettingsHooksNew}}new{{else}}{{.Webhook.ID}}{{end}}" method="post">
+		{{.CsrfTokenHtml}}
+		<div class="required field {{if .Err_BotToken}}error{{end}}">
+			<label for="bot_token">{{.i18n.Tr "repo.settings.bot_token"}}</label>
```
Comment (jonasfranz) [inline templates/repo/settings/webhook/telegram.tmpl:6]: Translation missing

**Review by jonasfranz — DISMISSED** [review]
Issues with HTML-titles won't be accepted by Telegram. I've got a 400 response: 

Request:
```json
{
  "text": "[\u003ca href=\"http://localhost:3000/SeriousOrg/protected\"\u003eSeriousOrg/protected\u003c/a\u003e] Issue opened: \u003ca href=\"http://localhost:3000/api/v1/repos/SeriousOrg/protected/issues/4\"\u003e#4 \u003cb\u003etest\u003c/b\u003e\u003c/a\u003e\n\n",
  "parse_mode": "HTML"
}
```

Response:
```json
{"ok":false,"error_code":400,"description":"Bad Request: can't parse entities: Expected end tag at byte offset 172"}
```

**Comment by techknowlogick** [comment]
Moving this to 1.8 so I can resolve above issues.

File: models/webhook_telegram.go (hunk 1 above)
Comment (techknowlogick) [reply models/webhook_telegram.go]: done

**Comment by techknowlogick** [comment]
Looking at [docs](https://core.telegram.org/bots/api#html-style) the only supported HTML tags are: `b`, `strong`, `i`, `em`, `a`, `code`, and `pre`. All `<`,`>` and `&` symbols  that are not a part of a tag or an HTML entity must be replaced with the corresponding HTML entities (`&lt;`, `&gt;`, and `&amp;` which along with `&quot;` are the only named HTML entities allowed). All numerical HTML entities are supported.

So I'm going to try to instead send HTML to the API, I will try to transform this into markdown, and send it that way.

File: models/webhook_telegram.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (adelowo) [inline models/webhook_telegram.go]: 2019

Comment (techknowlogick) [reply models/webhook_telegram.go]: done

**Comment by techknowlogick** [comment]
@lunny @jonasfranz I'm now passing all messages through the sanitizer, and messages are successfully sending.

**Review by lunny — APPROVED** [review]

**Review by sapk — APPROVED** [review]

**Comment by lafriks** [comment]
@jonasfranz please review

**Comment by wxiaoguang** [comment]
> Looking at [docs](https://core.telegram.org/bots/api#html-style) the only supported HTML tags are: `b`, `strong`, `i`, `em`, `a`, `code`, and `pre`. All `<`,`>` and `&` symbols that are not a part of a tag or an HTML entity must be replaced with the corresponding HTML entities (`&lt;`, `&gt;`, and `&amp;` which along with `&quot;` are the only named HTML entities allowed). All numerical HTML entities are supported.
> 
> So I'm going to try to instead send HTML to the API, I will try to transform this into markdown, and send it that way.

The root problem is that markdown shouldn't be mixed with HTML. The complete fix could be this:  Refactor webhook #31587

**Pull request merged**